File: contest/noi/NOI_2011_Road_Construction.java
package contest.noi;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class NOI_2011_Road_Construction {

	static BufferedReader br = new BufferedReader(new InputStreamReader(
			System.in));
	static PrintWriter ps = new PrintWriter(new BufferedWriter(
			new OutputStreamWriter(System.out)));
	static StringTokenizer st;
	static long[] size;
	static ArrayList<ArrayList<Integer>> adj;

	public static void main (String[] args) throws IOException {
		int n = readInt();
		size = new long[n];
		adj = new ArrayList<ArrayList<Integer>>();
		ArrayList<Edge> e = new ArrayList<Edge>();
		for (int x = 0; x < n; x++)
			adj.add(new ArrayList<Integer>());
		for (int x = 0; x < n - 1; x++) {
			int a = readInt() - 1;
			int b = readInt() - 1;
			int c = readInt();
			adj.get(a).add(b);
			adj.get(b).add(a);
			e.add(new Edge(a, b, c));
		}
		boolean[] v = new boolean[n];
		dfs(0, v);
		long cost = 0;
		for (Edge i : e) {
			long a = size[i.src];
			long b = size[i.dest];
			cost += i.cost * Math.abs((a > b ? ((n - b) - b) : ((n - a) - a)));
		}
		System.out.println(cost);
	}

	private static long dfs (int i, boolean[] v) {
		v[i] = true;
		int total = 0;
		for (Integer x : adj.get(i))
			if (!v[x])
				total += dfs(x, v);
		size[i] = total + 1;
		return size[i];
	}

	static class Edge {
		int src, dest, cost;

		Edge (int src, int dest, int cost) {
			this.src = src;
			this.dest = dest;
			this.cost = cost;
		}
	}

	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}

	static long readLong () throws IOException {
		return Long.parseLong(next());
	}

	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}

	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}

	static char readCharacter () throws IOException {
		return next().charAt(0);
	}

	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: src/contest/noi/NOI_2011_Road_Construction.cc
#include <bits/stdc++.h>

using namespace std;

struct edge{
    int src;
    int dest;
    int cost;
    edge(int src, int dest, int cost){
        this->src = src;
        this->dest = dest;
        this->cost = cost;
    }
};

int n;
long long size[1000000];
bool v[1000000];
vector<vector<int> > adj (1000000);
vector<edge> e;

static long long dfs(int i){
    v[i] = true;
    long long total = 0;
    for(int x = 0; x < adj[i].size(); x++){
        if(!v[adj[i][x]])
            total += dfs(adj[i][x]);
    }
    size[i] = total+1;
    return size[i];
}
int main(){
    scanf("%d",&n);
    for(int x = 0; x < n-1; x++){
        int a, b, c;
        scanf("%d%d%d",&a,&b,&c);
        a--;
        b--;
        adj[a].push_back(b);
        adj[b].push_back(a);
        e.push_back(edge(a,b,c));
    }
    dfs(0);
    long long cost = 0;
    for(int x = 0; x < e.size(); x++){

        long long a = size[e[x].src];
        long long b = size[e[x].dest];
        cost += e[x].cost * abs((a > b) ? (n-2*b):(n-2*a));
    }
    printf("%lld", cost);
    return 0;
}
